### PR TITLE
Feature/workspace grid view

### PR DESCRIPTION
**Pull request by ResearchDaniel**
Added a workspace gridview to welcome widget and made some visual changes related to it. 
It is possible to switch between the grid/tree views and the preference of which to show is stored.

Both the treeview and the gridview now shows the first canvas image, if available, as a workspace icon.
Skärmavbild 2022-01-14 kl  21 12 12

### CODE REVIEW

**Comment by ResearchDaniel** [comment]
@martinfalk I would like to make the changelog reserve less space. Tried various things, like QSizePolicy, minimum width and so on, but never figured out how. If you know how to do it I would be happy for some pointers...

**Comment by ResearchDaniel** [comment]
Now also tested on a high-res screen. 
image
image

**Comment by petersteneteg** [comment]
Im looking into this. I like the idea of the icons. But I think the overall layout of the widget got worse. The visual balance get thrown off I think. Having a center pane and a huge left pane seems off. I expect the center pane to be the dominating one. Not the other way around. I also don't agree the the changeling should be that tiny. 

I also looked at the implementation of the grid view and was quite confused. Are you layouting the grid manually using QVBoxklayouts and QListViews for sub trees of the model? is there no cleaner way to do this? 
Also are you loading __all__ the annotations upfront before each show of the widget?

**Comment by petersteneteg** [comment]
Only tried this in debug so far, but then my computer does a 5-10 sec beachball to open the dialog...

**Comment by ResearchDaniel** [comment]
Did most of the work on the laptop, which does not have as much screen space. So, the layout can be optimized a bit.

> Having a center pane and a huge left pane seems off

Now that we have images as icons we might want to remove the center layout and just keep the buttons below. I don't think that the Network image and other images gives a huge added value. We could also include the other images in the layout below.

> I also don't agree the the changeling should be that tiny.
We could make it bigger. Before it consumed almost half of the screen which was way too much. I do not think that the changelog is very important compared to the workspaces, so it should not consume much space, but also not look crammed in 

> Are you layouting the grid manually using QVBoxklayouts and QListViews for sub trees of the model? is there no cleaner way to do this?

I struggled quite a bit on getting a nice solution for this. The short answer is no. One could also use QGraphicsScene, but then you would have to re-implement a lot of things that you get for free in the QListView 

> Also are you loading all the annotations upfront before each show of the widget?

Yes, I might try some threaded loading but have not gotten around to it. Worked well on in debug mode on my mac. A bit worse in debug mode on my windows PC

**Comment by ResearchDaniel** [comment]
> Only tried this in debug so far, but then my computer does a 5-10 sec beachball to open the dialog...

Ok, then we need to fix it

**Comment by petersteneteg** [comment]
Im in the process of revising this

**Comment by petersteneteg** [comment]
@ResearchDaniel you might want to try out my changes to see If I accidentally broke anything you did.

**Comment by petersteneteg** [comment]
@martinfalk now you will have to review...

**Comment by petersteneteg** [comment]
needs to be rebase on #1238

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 Here we document changes that affect the public API or changes that needs to be communicated to other developers. 
 
+## 2022-02-01 Icon Grid View
+The welcome widget now has an icon grid view with previews of the first canvas in each network.
```
Comment (martinfalk) [inline CHANGELOG.md:4]: Feature request: make it possible in the network to make a Canvas processor the first one.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 Here we document changes that affect the public API or changes that needs to be communicated to other developers. 
 
+## 2022-02-01 Icon Grid View
+The welcome widget now has an icon grid view with previews of the first canvas in each network. 
+The workspace and processor search fields gained new features. One can for example search for all networks with a `Volume Raycaster` processor that was create by `Martin` like this: `processors:\"Volume Raycaster\" author:Martin"`. The tooltip for each search field has more details.
```
Comment (martinfalk) [inline CHANGELOG.md]:
```suggestion
The workspace and processor search fields gained new features. One can for example search for all networks with a `Volume Raycaster` processor that was created by `Martin` like this: `processors:\"Volume Raycaster\" author:Martin"`. The tooltip for each search field has more details. 
```

File: include/inviwo/qt/editor/workspacemodelroles.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+}
+inline bool operator!=(WorkspaceModelType t, const QVariant& v) { return operator!=(v, t); }
+
+namespace utilqt {
+
+template <typename Enum>
+QVariant getData(const QModelIndex& index, Enum e) {
+    return index.data(static_cast<std::underlying_type_t<Enum>>(e));
+}
+inline QVariant getData(const QModelIndex& index, Qt::ItemDataRole e) { return index.data(e); }
```
Comment (martinfalk) [inline include/inviwo/qt/editor/workspacemodelroles.h]:
```suggestion
inline QVariant getData(const QModelIndex& index, Qt::ItemDataRole r) { return index.data(r); }
```

File: src/core/CMakeLists.txt [hunk 4]
```diff
@@ -8,6 +8,7 @@ set(HEADER_FILES
     ${IVW_INCLUDE_DIR}/inviwo/core/algorithm/cubeplaneintersection.h
     ${IVW_INCLUDE_DIR}/inviwo/core/algorithm/linearmap.h
     ${IVW_INCLUDE_DIR}/inviwo/core/algorithm/permutations.h
+    ${IVW_INCLUDE_DIR}/inviwo/core/algorithm/seachdsl.h
```
Comment (martinfalk) [inline src/core/CMakeLists.txt]: typo in file name
```suggestion
    ${IVW_INCLUDE_DIR}/inviwo/core/algorithm/searchdsl.h
```

File: src/core/CMakeLists.txt [hunk 5]
```diff
@@ -386,6 +387,7 @@ set(SOURCE_FILES
     algorithm/cubeplaneintersection.cpp
     algorithm/linearmap.cpp
     algorithm/permutations.cpp
+    algorithm/seachdsl.cpp
```
Comment (martinfalk) [inline src/core/CMakeLists.txt]: typo in file name

File: include/inviwo/core/algorithm/seachdsl.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
+/*********************************************************************************
```
Comment (martinfalk) [inline include/inviwo/core/algorithm/seachdsl.h]: rename to `searchdsl.h`

File: src/core/algorithm/seachdsl.cpp [hunk 7]
```diff
@@ -0,0 +1,32 @@
+/*********************************************************************************
```
Comment (martinfalk) [inline src/core/algorithm/seachdsl.cpp]: rename to `searchdsl.cpp`

File: src/qt/editor/inviwomainwindow.cpp [hunk 8]
```diff
@@ -1337,6 +1306,47 @@ void InviwoMainWindow::visibilityModeChangedInSettings() {
     updateWindowTitle();
 }
 
+WelcomeWidget* inviwo::InviwoMainWindow::getWelcomeWidget() {
+    // The welcome widget is rather expensive to keep around, so do not create it unnecessarily.
+    // Tested on Windows, where 50 MB was released upon deletion.
```
Comment (martinfalk) [inline src/qt/editor/inviwomainwindow.cpp]:
```suggestion
    // Tested on Windows, where 50 MB were released upon deletion.
```

File: src/qt/editor/processorlistwidget.cpp [hunk 9]
```diff
@@ -142,9 +142,41 @@ void ProcessorTree::showContextMenu(const QPoint& p) {
...
+    return [mptr](const ProcessorFactoryObject& p, std::string_view str) -> bool {
+        return strMatch(std::invoke(mptr, p), str);
+    };
+};
+
 ProcessorTreeWidget::ProcessorTreeWidget(InviwoMainWindow* parent, HelpWidget* helpWidget)
     : InviwoDockWidget(tr("Processors"), parent, "ProcessorTreeWidget")
     , app_{parent->getInviwoApplication()}
+    , dsl_{{
+          {"identifier", "i", "processor class identifier", true,
```
Comment (martinfalk) [inline src/qt/editor/processorlistwidget.cpp]: Possibility to query the module as well?

File: src/qt/editor/welcomewidget.cpp [hunk 10]
```diff
@@ -444,89 +616,109 @@ void WelcomeWidget::updateDetails(const QString& filename) {
...
-    replaceInString(titleStr, "-", " ");
-    replaceInString(titleStr, "_", " ");
+    const auto createdStr = utilqt::fromQString(info.birthTime().toString(dateformat));
+    const auto modifiedStr = utilqt::fromQString(info.lastModified().toString(dateformat));
+    const auto titleStr = [&]() {
+        auto str = (annotations && !annotations->getTitle().empty())
+                       ? annotations->getTitle()
+                       : utilqt::fromQString(info.completeBaseName());
+        if (!str.empty()) str[0] = static_cast<char>(std::toupper(str[0]));
+        replaceInString(str, "-", " ");
```
Comment (martinfalk) [inline src/qt/editor/welcomewidget.cpp]: always a good idea? Can be rather annoying when the custom annotations title gets messed with.

File: src/qt/editor/welcomewidget.cpp [hunk 11]
```diff
@@ -444,89 +616,109 @@ void WelcomeWidget::updateDetails(const QString& filename) {
...
+        addRow("Description", description);
+    } else {
         body.append("h3", "Workspace file could not be opened!",
                     {{"style", "color:firebrick;font-weight:600;"}});
     }
+
     auto content = body.append("span");
-    auto addImage = [&content](auto item) {
-        const int fixedImgHeight = 256;
+    const int maxImageSize = utilqt::emToPx(this, 32);
```
Comment (martinfalk) [inline src/qt/editor/welcomewidget.cpp:681]: Make it adjustable via global settings, default 32em.

**Review by martinfalk — CHANGES_REQUESTED** [review]
Nice work!

Some minor comments and suggestions.

File: CHANGELOG.md (hunk 1 above)
Comment (petersteneteg) [reply CHANGELOG.md:4]: You can set the primary canvas identifier in the workspace annotations now.

File: src/qt/editor/processorlistwidget.cpp (hunk 9 above)
Comment (petersteneteg) [reply src/qt/editor/processorlistwidget.cpp]: added

File: src/qt/editor/welcomewidget.cpp (hunk 11 above)
Comment (petersteneteg) [reply src/qt/editor/welcomewidget.cpp:681]: we don't have any qt settings. And I think is is better to make sure the default is good enough.

**Review by petersteneteg — APPROVED** [review]

**Review by martinfalk — APPROVED** [review]

**Pull request merged**